### PR TITLE
Added `estimate_KM.formula()` method

### LINKED ISSUE #379: Add `estimate_KM.formula()` method
The visualizations for time to event data in visR are amazing and will be used by many people, including those whose data does not follow CDISC conventions. We can do a couple of things to improve their experience, while not taking away from the experience of those working with CDISC data.

One **HUGE** difference between most statistical coding and CDISC is the CNSR column: it's the opposite of what many will expect. TO aid these users, adding a formula method to `estimate_KM()` will allow non-CDISC users to use familiar formula and data syntax, while the CDISC users can continue to utilize the helpful CDISC defaults.

The `estimate_KM.formula()` method will simply parse the formula, and pass the elements to `estamate_KM(data, AVAL, CNSR, strata)`; a simple function adding great usability to many more users.

### PR DESCRIPTION
**Pull request by ddsjoberg**
**What changes are proposed in this pull request?**
* Converted `estimate_KM()` to a generic and added S3 methods `estimate_KM.data.frame()` and `estimate_KM.formula()`. The data frame method is optimized for CDISC data, and the formula uses the standard survival syntax. 

The `estimate_KM()` formula method is now marked as experimental and is not implemented for other functions at this time.

**Did you include unit tests for the proposed change/bug fix (https://testthat.r-lib.org/)?**
Yes

**If there is an GitHub issue associated with this pull request, please provide link.**
closes #379 

--------------------------------------------------------------------------------

Checklist for PR reviewer

- [ ] PR branch has pulled the most recent updates from main branch. Ensure the pull request branch and your local version match and both have the latest updates from the main branch.
- [ ] If a new function was added, function should be included in `_pkgdown.yml`
- [ ] If a bug was fixed, a unit test was added for the bug check
- [ ] Run `pkgdown::build_site()`. Check the R console for errors, and review the rendered website.
- [ ] Code coverage is suitable for any new functions/features. Review coverage with `withr::with_envvar(new = c("NOT_CRAN" = "true"), covr::report())`. Before you run, begin a fresh R session without any packages loaded. 
- [ ] R CMD Check runs without errors, warnings, and notes
- [ ] `usethis::use_spell_check()` runs with no spelling errors in documentation
- [ ] Has `NEWS.md` been updated with the changes from this pull request under the heading indicating the latest version. If there is an issue associated with the pull request, reference it in parentheses at the end update (see `NEWS.md` for examples).
- [ ] Has the version number been incremented using `usethis::use_version(which = "dev")` 
- [ ] Approve Pull Request
- [ ] Merge the PR. Please use "Squash and merge".

### CODE REVIEW

File: R/estimate_KM.R [hunk 1]
```diff
@@ -65,27 +69,27 @@
...
   dots <- rlang::dots_list(...)
+  if ("formula" %in% names(dots)) {
+    formula <- dots[["formula"]]
+    dots[["formula"]] <- NULL
+    message("To use the `estimate_KM(formula=)` argument, place it in the first position.")
+    return(rlang::inject(estimate_KM.formula(formula = formula, data = data, !!!dots)))
+  }
 
   # Validate argument inputs ---------------------------------------------------
-
```
Comment (SHAESEN2) [inline R/estimate_KM.R:82]: Why is this removed? Is it automatically handled by the method? If yes the testing needs to be updated.

File: tests/testthat/test-get_pvalue.R [hunk 2]
```diff
@@ -240,7 +240,7 @@ testthat::test_that("T5.1 P-values are accurate when a filtered data frame is pi
     ) %>%
     {pchisq(.$chisq, length(.$n) - 1, lower.tail = FALSE)} %>%
     visR:::.pvalformat()
-  testthat::expect_equal(survfit_p, survdiff_p)
```
Comment (SHAESEN2) [inline tests/testthat/test-get_pvalue.R:243]: Use prefixes everywhere.

File: tests/testthat/test-estimate_KM.R [hunk 3]
```diff
@@ -43,6 +43,9 @@
 #' T7.2 The function prefixes the function call with survival
 #' T8. Piped datasets still return accurate results
 #' T8.1 Piped datasets still return accurate results
+#' T9. The user can specify formula argument
```
Comment (SHAESEN2) [inline tests/testthat/test-estimate_KM.R:46]: Are all new requirements and code tested?

File: tests/testthat/test-estimate_KM.R [hunk 4]
```diff
@@ -43,6 +43,9 @@
...
+#' T9.1 The formula method returns the same results and the data method.
+#' T9.2 The formula method triggers error messages.
```
Comment (SHAESEN2) [inline tests/testthat/test-estimate_KM.R:48]: this is too vague

**Review by SHAESEN2 — CHANGES_REQUESTED** [review]
Reviewing this properly would take me some time. At a first look, it looks great but too complex for what it actually is doing. I feel like this should be part of a next CRAN release so we can properly look into this. We can discuss during our next meeting.

File: R/estimate_KM.R (hunk 1 above)
Comment (ddsjoberg) [reply R/estimate_KM.R:82]: Yes, auto-handled by the S3 method.

File: tests/testthat/test-estimate_KM.R (hunk 4 above)
Comment (ddsjoberg) [reply tests/testthat/test-estimate_KM.R:48]: Feel free to update the text to your preference

**Comment by ddsjoberg** [comment]
> Reviewing this properly would take me some time. At a first look, it looks great but too complex for what it actually is doing. I feel like this should be part of a next CRAN release so we can properly look into this. We can discuss during our next meeting.

We can ask @bailliem to assist with the review if you don't have time at the moment.

Regarding , "too complex for what it actually is doing", I am not sure what you mean. A simpler approach is to add a formula argument and avoid generics/S3 methods (still my preference), but I think you preferred the S3 method when we discussed.

File: R/estimate_KM.R [hunk 5]
```diff
@@ -15,24 +12,31 @@
 #'      \item{A two-sided pointwise 0.95 confidence interval is estimated using a log transformation (conf.type = "log").}
 #'    }
 #'
+#'    Function has two S3 methods for data frames and formulas. The first is optimized for CDISC data, using default naming conventions,
```
[the remark this answers is not in the document]
Comment (ddsjoberg) [reply R/estimate_KM.R:15]: updated

File: tests/testthat/test-estimate_KM.R (hunk 3 above)
Comment (ddsjoberg) [reply tests/testthat/test-estimate_KM.R:46]: yes

**Comment by ddsjoberg** [comment]
@bailliem @SHAESEN2 

In @SHAESEN2 's review, I think he accurately points out that the S3 methods are overly complex given the added functionality, and I agree with him. We should opt for simply adding the `estimate_KM(formula=)` argument to the function. I put together a new pull request (https://github.com/openpharma/visR/pull/390) based on the review comments that adds the new argument rather than implementing generics/S3 methods. The new implementation is much simpler and will be easier to maintain.

I am going to close this PR for now, and we can discuss on Tuesday.
